pxr/imaging/hgiPresent/vulkan.cpp

```diff
@@ -0,0 +1,756 @@
+    } else if (surfaceCapabilities.supportedCompositeAlpha &
+        VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR) {
+        _alphaComposition = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
+    } else {
+        TF_WARN(
+            "Vulkan opaque or inherit alpha composition unsupported: presentation is disabled");
+        return;
+    }
+
+    _presentMode = VK_PRESENT_MODE_FIFO_KHR;
+    if (!_params.wantVsync) {
+        uint32_t presentModeCount{};
+        HGIVULKAN_VERIFY_VK_RESULT(
+            vkGetPhysicalDeviceSurfacePresentModesKHR(physicalDevice, _surface,
+                & presentModeCount, nullptr));
+        std::vector<VkPresentModeKHR> presentModes{presentModeCount};
+        HGIVULKAN_VERIFY_VK_RESULT(
+            vkGetPhysicalDeviceSurfacePresentModesKHR(physicalDevice, _surface,
+                & presentModeCount, presentModes.data()));
+        if (std::find(presentModes.begin(), presentModes.end(),
+            VK_PRESENT_MODE_IMMEDIATE_KHR) != presentModes.end()) {
+            _presentMode = VK_PRESENT_MODE_IMMEDIATE_KHR;
+        }
+    }
+
+    if (!(surfaceCapabilities.supportedTransforms &
+        VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR)) {
+        // According to the Vulkan capability database,
+        // 100% of drivers support this so we should never get here.
+        TF_WARN(
+            "Vulkan surface transform is unsupported: presentation is disabled");
+        return;
+    }
+
+    // Note: while it's possible to get the surface size from
+    // surfaceCapabilities.currentExtent, it's not guaranteed to be
+    // available, so we still need to rely on the window size coming
+    // from the present task.
+    _swapchainExtent.width = std::clamp(static_cast<uint32_t>(_srcTextureSize[0]),
+        surfaceCapabilities.minImageExtent.width,
+        surfaceCapabilities.maxImageExtent.width);
+    _swapchainExtent.height = std::clamp(static_cast<uint32_t>(_srcTextureSize[1]),
+        surfaceCapabilities.minImageExtent.height,
+        surfaceCapabilities.maxImageExtent.height);
+
+    const auto minComponents = HgiGetComponentCount(
+        _params.preferredSurfaceFormat);
+    const auto preferredColorVolume = _GetColorVolume(
+        HgiVulkanConversions::GetFormat(_params.preferredSurfaceFormat));
+
+    uint32_t formatCount{};
+    HGIVULKAN_VERIFY_VK_RESULT(
+        vkGetPhysicalDeviceSurfaceFormatsKHR(physicalDevice, _surface, &
+            formatCount, nullptr));
+    std::vector<VkSurfaceFormatKHR> formats{formatCount};
+    HGIVULKAN_VERIFY_VK_RESULT(
+        vkGetPhysicalDeviceSurfaceFormatsKHR(physicalDevice, _surface, &
+            formatCount, formats.data()));
+
+    using CandidateFormat = std::pair<VkSurfaceFormatKHR, _ColorVolume::Match>;
+    std::vector<CandidateFormat> candidateFormats;
+    for (const auto formatAndColorSpace : formats) {
+        const auto [format, colorSpace] = formatAndColorSpace;
+        if (_VkColorSpaceEnumToName(colorSpace) != _params.surfaceColorSpace) {
```

Comment:
similarly to the metal case, although the hardware can help us with some conversions, it can't help with all. So I think there should also be a use of the GfColorSpace matrices and coefficients in a shader to help with those cases. VK doesn't know about ACES for example.

Reply:
How would we know the output of the colour correction task is, for example, ACES? As far as I can tell, only linear extended sRGB, non linear (extended?) sRGB, and some user defined OCIO spaces are supported. The user defined colour space could be ACES, but the strings used by OCIO aren't standardized; they're only IDs to the actual transformation in a config file. At least that's what I've been told.

Reply:
This comment followed on the previous one, in the case that we are intending to do color space conversions. As discussed there, currently the target is just versus what the system currently does in linear rec709, so the question of supporting conversions here is out of scope :)